### ISSUE TITLE
Error: `headers` was called outside a request scope.

### ISSUE THREAD

**Issue by BayuPrasetyaUtomo** [get-convex/convex-auth#120]
The issue appears when using SSA in Next JS

## Runtime: 
Bun version 1.1.34

## Package version:
```json
  "convex": "^1.17.0"
  "@auth/core": "0.36.0"
  "@convex-dev/auth": "^0.0.74"
  "next": "^14.2.17"
  "react": "^18.3.1"
  "react-dom": "^18.3.1"
```

## Error message:
```
 ✓ Compiled middleware in 10ms
 ✓ Compiled in 5ms
 ⨯ 37 |     return f;
38 | }
39 | function getServerError(error, type) {
40 |     let n;
41 |     try {
42 |         throw new Error(error.message);
                       ^
error: `headers` was called outside a request scope. Read more: https://nextjs.org/docs/messages/next-dynamic-api-wrong-context
```

The issue started when adding these lines to the middleware:
```ts
// middleware.ts
import { convexAuthNextjsMiddleware } from "@convex-dev/auth/nextjs/server";
 
export default convexAuthNextjsMiddleware();
 
export const config = {
  // The following matcher runs middleware on all routes
  // except static assets.
  matcher: ["/((?!.*\\..*|_next).*)", "/", "/(api|trpc)(.*)"],
};
```
The issue dissapear as soon as the `convexAuthNextjsMiddleware` was removed from the middleware and changed to Next js default middleware.
```ts
// middleware.ts
const middleware = () => {

}

export const config = {
  // The following matcher runs middleware on all routes
  // except static assets.
  matcher: ["/((?!.*\\..*|_next).*)", "/", "/(api|trpc)(.*)"],
};

export default middleware;
```

<details>
  <summary><h2>Minimum folder structure:</h2></summary>

`node_modules` `.next` and `public` folder has been excluded
```text
  📦project
 ┣ 📂convex
 ┃ ┣ 📂_generated
 ┃ ┃ ┣ 📜api.d.ts
 ┃ ┃ ┣ 📜api.js
 ┃ ┃ ┣ 📜dataModel.d.ts
 ┃ ┃ ┣ 📜server.d.ts
 ┃ ┃ ┗ 📜server.js
 ┃ ┣ 📜auth.config.ts
 ┃ ┣ 📜auth.ts
 ┃ ┣ 📜http.ts
 ┃ ┣ 📜schema.ts
 ┃ ┗ 📜tsconfig.json
 ┣ 📂src
 ┃ ┣ 📂app
 ┃ ┃ ┣ 📜layout.tsx
 ┃ ┃ ┗ 📜page.tsx
 ┃ ┣ 📂components
 ┃ ┃ ┗ 📂provider
 ┃ ┃ ┃ ┗ 📜auth-client.tsx
 ┃ ┣ 📂styles
 ┃ ┃ ┗ 📜globals.css
 ┃ ┗ 📜middleware.ts
 ┣ 📜.env.local
 ┣ 📜.eslintrc.cjs
 ┣ 📜.gitignore
 ┣ 📜README.md
 ┣ 📜bun.lockb
 ┣ 📜next-env.d.ts
 ┣ 📜next.config.js
 ┣ 📜package.json
 ┣ 📜postcss.config.js
 ┣ 📜prettier.config.js
 ┣ 📜tailwind.config.ts
 ┗ 📜tsconfig.json
 ```
 </details>
 
<details>
  <summary><h2>Layout:</h2></summary>
  
```tsx
// layout.tsx
import "@/styles/globals.css";
import { ConvexClientProvider } from "@/components/provider/auth-client";

export default function RootLayout({
    children,
  }: Readonly<{ children: React.ReactNode }>) {
    return (
        <html lang="en">
          <body>
            <ConvexClientProvider>{children}</ConvexClientProvider>
          </body>
        </html>
    );
  }
```
</details>

<details>
  <summary><h2>Home</h2></summary>

```tsx
// src/app/page.tsx
"use client";
import { useAuthActions } from "@convex-dev/auth/react";
import { Authenticated, AuthLoading, Unauthenticated } from "convex/react";

export default function HomePage() {
  const { signIn, signOut } = useAuthActions();
  return (
    <main className="flex min-h-screen flex-col items-center justify-center bg-gradient-to-b from-[#2e026d] to-[#15162c] text-white">
      <AuthLoading>Loading...</AuthLoading>

      <Unauthenticated>
        <button onClick={() => void signIn("github")}>Sign in</button>
      </Unauthenticated>
      <Authenticated>
        Hello There!
        <button onClick={() => void signOut()}>Sign out</button>
      </Authenticated>
    </main>
  );
}
</details>

**Comment by timarnoldev** [comment]
Same problem here

**Comment by HarunKilic** [comment]
Did you figure this out?

**Comment by dandanthedev** [comment]
+1

**Comment by dandanthedev** [comment]
temporary fix is to use "App router without SSA" here: https://labs.convex.dev/auth/setup

tldr: remove middleware.ts and ConvexAuthNextjsServerProvider and add the following to your layout
```tsx
import { ConvexAuthProvider } from "@convex-dev/auth/react";
const convex = new ConvexReactClient(process.env.NEXT_PUBLIC_CONVEX_URL!);

export default function MyLayout(){
  return <ConvexAuthProvider client={convex}>
               //rest of your layout junk
          </ConvexAuthProvider>;
}
```


issue is this disallows any server-side rendering that needs auth. solved it for me tho

**Comment by jankc** [comment]
I had the same issue until I migrated to Next v16 and switched from middleware to proxy. I basically just renamed `middleware.ts` to `proxy.ts` and changed the export line: 
```ts
export const proxy = convexAuthNextjsMiddleware();
```
It works locally and even in Vercel's newly supported Bun runtime (`"bunVersion": "1.x"` in vercel.json).
However, I'm not sure if it does not violate the Vercel rules for `proxy.ts` not to contain heavy auth logic.
